enableManifestClasspath should also affect linux and init.sh start scripts

jfosco opened the pull request:
You can now define a customClasspath for the distribution which will
override the expansion of all *.jar files in the runtime configuration.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/palantir/gradle-java-distribution/128)

<!-- Reviewable:end -->

uschi2000 commented on src/test/groovy/com/palantir/gradle/javadist/JavaDistributionPluginTests.groovy:
```diff
@@ -309,6 +310,47 @@ class JavaDistributionPluginTests extends GradleTestSpec {
+        expectedStaticConfig.setConfigType("java")
+        expectedStaticConfig.setJavaHome("")
+        expectedStaticConfig.setMainClass("test.Test")
+        expectedStaticConfig.setClasspath(['service/lib/manager.jar', 'service/lib/config.jar'])
```
can probably worry about this field only in the assertion

uschi2000 commented on src/main/groovy/com/palantir/gradle/javadist/tasks/LaunchConfigTask.groovy:
```diff
@@ -65,8 +67,12 @@ class LaunchConfigTask extends BaseTask {
         config.args = args
-        config.classpath = relativizeToServiceLibDirectory(
-                project.tasks[JavaPlugin.JAR_TASK_NAME].outputs.files + project.configurations.runtime)
+        if (distributionExtension().getCustomClasspath().size() == 0) {
```
isEmpty

uschi2000 reviewed (changes requested):
(1) need to update readme
(2) what's the use case for this? the implementation seems fine module some nitpicks, but I don't quite understand the rationale.

uschi2000 commented:
further, we should fix the classpath at compile time; so it shouldn't contain `*` and other run-time -evaluated things like `../` etc

uschi2000 commented on src/main/groovy/com/palantir/gradle/javadist/DistributionExtension.groovy:
```diff
@@ -62,6 +65,16 @@ class DistributionExtension {
     }
 
+    public void customClasspath(String... customClasspath) {
+        assert customClasspath.every { isAllowedClasspathChar(it) }
```
should give more useful error of the form "Found foo, expected bar"

jfosco replied:
If I use the custom classpath to filter against the list of jars originally used in the LaunchConfigTask, do I still need to check for special characters?

uschi2000 replied:
IMO, it's fine to just check the custom ones

uschi2000 commented on readme.md:
```diff
@@ -62,6 +62,8 @@ The `distribution` block offers the following options:
  * (optional) `checkArgs` a list of arguments to supply to the monitoring script, if omitted,
    no monitoring script will be generated.
  * (optional) `defaultJvmOpts` a list of default JVM options to set on the program.
+ * (optional) `customClasspath` a list of jars to put on the classpath instead of all jars in the
```
this is not quite complete enough. at the minimum, must say that these are file names relative to some path, that they must not contain certain characters, something about the classpath order

uschi2000 commented on src/test/groovy/com/palantir/gradle/javadist/DistributionExtensionTest.groovy:
```diff
@@ -59,12 +63,41 @@ class DistributionExtensionTest extends Specification {
+
+        when:
+        ext.with {
+            setCustomClasspath(['*', '..', ':', '/'])
```
every single one of them should throw, right?

uschi2000 commented on src/test/groovy/com/palantir/gradle/javadist/JavaDistributionPluginTests.groovy:
```diff
@@ -309,6 +310,27 @@ class JavaDistributionPluginTests extends GradleTestSpec {
+        file('src/main/java/test/Test.java') << "package test;\npublic class Test {}"
+
+        when:
+        runSuccessfully(':build', ':distTar', ':untar')
```
i actually believe this should fail if the classpath files are not readable

jfosco replied:
Good point. I changed it to require that the customClasspath jars have to match actual dependency jars.

uschi2000 commented:
I still think this is a little convoluted, sorry. (I should have been more specific here from the start.) The default behaviour is to to use the concatenation of the jar and the runtime dependencies. I would propose to make provide an optional `FileCollection classpath` configuration point such that: if it is set, then this FileCollection becomes the classpath, else the default is the classpath (see above).

---

Review status: 0 of 5 files reviewed at latest revision, 1 unresolved discussion.

---

*[a discussion](https://reviewable.io:443/reviews/palantir/gradle-java-distribution/128#-KUecffZPyIUEj0gNIo6:-KUecffZPyIUEj0gNIo7:b-c6aubj) (no related file):*
Further, the ManifestClasspathJarTask must be updated accordingly, so this is likely a slightly bigger refactor.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/gradle-java-distribution/128#-:-KUebVYXP3vjxwoW4rEj:boazdqi)*
<!-- Sent from Reviewable.io -->

Commit:
enableManifestClasspath for linux and init.sh start scripts

If your java application has a long classpath you can surpass some linux
proc command length storage limits (4096 bytes) causing init.sh to fail.
You can now fix this by setting enableManifestClasspath

uschi2000 commented:
So this PR is now about enabling the classpath jar, rather than about allowing to customize the classpath, right?

---

Review status: 0 of 5 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/gradle-java-distribution/128#-:-KV2iyVxpcss5IU9_1kN:b9o7vqr)*
<!-- Sent from Reviewable.io -->

uschi2000 commented:
Review status: 0 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/test/groovy/com/palantir/gradle/javadist/JavaDistributionPluginTests.groovy, line 338 at r3](https://reviewable.io:443/reviews/palantir/gradle-java-distribution/128#-KV2lBXmUkACP4ussLyq:-KV2lBXnHeEe2uFdyC0J:b-lu6uj7) ([raw file](https://github.com/palantir/gradle-java-distribution/blob/561434821bd8cd74a7bf3f95ad9f9697eee89433/src/test/groovy/com/palantir/gradle/javadist/JavaDistributionPluginTests.groovy#L338)):*
> ```Groovy
>     }
> 
>     def 'produces manifest-classpath jar and windows start script with no classpath length limitations'() {
> ```

stale test name

---

*[src/test/groovy/com/palantir/gradle/javadist/JavaDistributionPluginTests.groovy, line 362 at r3](https://reviewable.io:443/reviews/palantir/gradle-java-distribution/128#-KV2lEcfuICOMRZtUXwT:-KV2lEcfuICOMRZtUXwU:b3ofc27) ([raw file](https://github.com/palantir/gradle-java-distribution/blob/561434821bd8cd74a7bf3f95ad9f9697eee89433/src/test/groovy/com/palantir/gradle/javadist/JavaDistributionPluginTests.groovy#L362)):*
> ```Groovy
>         String launcher = file('dist/service-name-0.1/service/bin/launcher-static.yml', projectDir).text
>         !launcher.contains("external.jar")
>     }
> ```

would like to see an end-to-end test for launching the service through the unix classpath jar

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/gradle-java-distribution/128)*
<!-- Sent from Reviewable.io -->

uschi2000 commented:
Reviewed 5 of 7 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/gradle-java-distribution/128)*
<!-- Sent from Reviewable.io -->

jfosco commented:
Done, but producing a manifest classpath jar is busted until you merge in YY's commit fixing how the plugin reaches into the gradle object model

uschi2000 commented:
Closing since there wasn't much  movement. Feel free to reopen if this is still an issue.
